subprojects/model-core/src/main/java/org/gradle/api/internal/provider/FilteringProvider.java

```diff
@@ -0,0 +1,101 @@
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import org.gradle.api.InvalidUserCodeException;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.function.Predicate;
+
+/**
+ * A filtering provider that uses a predicate to filter the value of another provider.
+ **/
+public class FilteringProvider<T> extends AbstractMinimalProvider<T> {
+
+    protected final Class<T> type;
+    protected final ProviderInternal<? extends T> provider;
+    protected final Predicate<? super T> predicate;
+
+    public FilteringProvider(
+        @Nullable Class<T> type,
+        ProviderInternal<? extends T> provider,
+        Predicate<? super T> predicate
+    ) {
+        this.type = type;
+        this.predicate = predicate;
+        this.provider = provider;
+    }
+
+    @Nullable
+    @Override
+    public Class<T> getType() {
+        return type;
+    }
+
+    @Override
+    public ValueProducer getProducer() {
+        return provider.getProducer();
+    }
+
+    @Override
+    public ExecutionTimeValue<? extends T> calculateExecutionTimeValue() {
+        ExecutionTimeValue<? extends T> value = provider.calculateExecutionTimeValue();
+        if (value.hasChangingContent()) {
+            // Need the value contents in order to transform it to produce the value of this provider,
+            // so if the value or its contents are built by tasks, the value of this provider is also built by tasks
+            return ExecutionTimeValue.changingValue(new FilteringProvider<>(type, value.toProvider(), predicate));
+        }
+
+        return ExecutionTimeValue.value(filterValue(value.toValue()));
+    }
+
+    @Override
+    protected Value<? extends T> calculateOwnValue(ValueConsumer consumer) {
+        beforeRead();
+        Value<? extends T> value = provider.calculateValue(consumer);
+        if (value.isMissing()) {
+            return value.asType();
+        }
+        return filterValue(value);
+    }
+
+    @Nonnull
+    protected Value<? extends T> filterValue(Value<? extends T> value) {
+        T unpackedValue = value.getWithoutSideEffect();
+        if (predicate.test(unpackedValue)) {
+            return value;
+        } else {
+            return Value.missing();
+        }
+    }
+
+    protected void beforeRead() {
+        provider.getProducer().visitContentProducerTasks(producer -> {
```

Comment:
🤔 Why does the filtering provider need to check the state of the delegate, wouldn't the delegate already check its own state?

Reply:
If I remove `beforeRead` from `FilteringProvider`, then the tests in `FilteringProviderTest` should fail with a different exception, but that's not the case. As discussed privately, the refactoring in this area is to be addressed in a separate issue.

Reply:
https://github.com/gradle/gradle/issues/26078